### ISSUE TITLE
Lower custom format score after rename causes redownloading loop of same release.

### ISSUE THREAD

**Issue by mrbruins** [Dictionarry-Hub/profilarr#140]
Over the past 3 weeks, for some TV shows, Sonarr has been downloading the exact same episode/release/file for over 25 times. See screenshot below:

<img width="1483" alt="image" src="https://github.com/user-attachments/assets/1454e40c-ca8b-472c-8f42-b67cc9719a51" />

For this show, I've narrowed this issue down to the custom format parser not detecting the `Peacock TV` custom format after renaming. 

The download decision is made based on the following name:
`The Day of the Jackal S01 REPACK 2160p PCOK WEB-DL DDP5 1 DV HDR H 265-FLUX`, 
which has the following custom formats: `2160p` `DD+` `Dolby Vision` `h265 (4k)` `HDR10` `Peacock TV`,
and a custom format score of 120.

Then after the file is renamed to:
`The Day of the Jackal (2024) [tvdbid-426866]/The Day of the Jackal (2024) - S01E07 - Episode 7 [Peacock TV WEBDL-2160p Proper][DV HDR10][EAC3 5.1][x265]-FLUX.mkv`,
which has the following custom formats: `2160p` `DD+` `Dolby Vision` `h265 (4k)` `HDR10` `x265 (Web)`,
and a custom format score of 100. 

I'm using the `2160p Optimal` quality profile, which assigns a score of 20 to the `Peacock TV` custom format. This custom format is replaced by `x265 (Web)` after the file rename, but this format has a score of 0 in the `2160p Optimal` profile. This is due to the release title condition `x265` in the `Peacock TV` custom format. 

I'm using the following Standard Episode Naming Format:
`{Series TitleYear} - S{season:00}E{episode:00} - {Episode CleanTitle} [{Custom Formats }{Quality Full}]{[MediaInfo VideoDynamicRangeType]}{[Mediainfo AudioCodec}{ Mediainfo AudioChannels]}{[MediaInfo VideoCodec]}{-Release Group}`

Is this reduction of the overall custom format score after file renaming intended behaviour or is this a bug in the `Peacock TV` custom format and/or the `2160p Optimal` profile? As a workaround, I've set the `Minimum Custom Format Score Increment` to 21 as a workaround for now.

**Comment by santiagosayshey** [comment]
hey there @mrbruins, thanks for bringing this up. I believe this is a known issue at the moment, see our discussion here:
https://discord.com/channels/1202375791556431892/1316075375616462989

The issue actually has nothing to do with file renames - the score given by the renamed file doesn't affect any future upgrades.

What is actually happening, is the *filename* doesn't include h265, which is different from the *torrent/nzb name* which does.  When you download the file, sonarr sees (h265, score = 120), but then when you import, the score drops because it can't find h265 anymore. This is why it keeps trying to download over and over again. 

Our working solution is imply h265 from filenames using specific patterns - includes mkv, has hevc/x265, etc. 

```
{
  "name": "h265 (Missing)",
  "includeCustomFormatWhenRenaming": false,
  "specifications": [
    {
      "name": "Release Group",
      "implementation": "ReleaseGroupSpecification",
      "negate": false,
      "required": true,
      "fields": {
        "value": "(?<=^|[\\s.-])SiGLA|FLUX\\b"
      }
    },
    {
      "name": "WEB-DL",
      "implementation": "SourceSpecification",
      "negate": false,
      "required": true,
      "fields": {
        "value": 3
      }
    },
    {
      "name": "Filename",
      "implementation": "ReleaseTitleSpecification",
      "negate": false,
      "required": true,
      "fields": {
        "value": ".mkv|.mp4"
      }
    },
    {
      "name": "x265 | HEVC",
      "implementation": "ReleaseTitleSpecification",
      "negate": false,
      "required": true,
      "fields": {
        "value": "^(?!.*(?i:remux))(?=.*(\\b[x]\\s?(\\.?265)\\b|HEVC|\\bDS4K\\b)).*$"
      }
    },
    {
      "name": "Not h265",
      "implementation": "ReleaseTitleSpecification",
      "negate": true,
      "required": false,
      "fields": {
        "value": "(?i)h\\s*\\.?\\s*265"
      }
    }
  ]
}

and set its score to whatever h265 was for your profile.

I'm guessing you might have grabbed a file that looks something like this? 

Torrent Name: `The Day of the Jackal S01 REPACK 2160p PCOK WEB-DL DDP5 1 DV HDR H 265-FLUX`
File Name: `The.Day.Of.The.Jackal.S01E07.REPACK.2160p.PCOK.WEB-DL.DDP5.1.DV.HEVC-FLUX.mkv`

Give the custom format above a try, see if it fixes the problem. If it doesn't, could you provide the filenames for the grabs that got upgraded?

Cheers
- Santiago

**Comment by mrbruins** [comment]
### TLDR;
1. Total CF score gets determined by the release title on the RSS feed.
2. The release name contains `PCOK` so matches the `Peacock TV` CF, while the release name does not contain `x265` (this is important!).
3. After importing, the substring `x265` is added because of the ffprobe metadata, which is part of the file renaming format.
4. Now the `x265` substring in the filename matches the negating condition in the `Peacock TV` CF, removing the CF from the file.
5. Since the `Peacock TV` CF has 20 points assigned to it in the quality profile, the total score of the release title is 20 points higher before download than after import.
6. This causes the upgrade loop.

Why is the `x265` negating condition part of the _network_ custom formats?

### Longer explanation
> The issue actually has nothing to do with file renames - the score given by the renamed file doesn't affect any future upgrades.

I'm not sure if that is the case, but even if so, the h265 custom formats can't be the issue, given that they all have a score of 0 in the `2160P Optimal` profile. The [documentation](https://wiki.servarr.com/sonarr/settings#custom-formats-2) shows that the filename after importing does matter:
>Release Title - This is a regular expression matched against the release title **and, after download, the filename on disk.**
>Note: Sonarr only considers text after the series title which means anything preceding the title is ignored.
>**Note that the matching text is needed to be included in the renamed file to ensure continual matching of the custom format after import.**

I downloaded this episode using SabNZBd.
NZB name: `The.Day.of.the.Jackal.S01E07.Episode.7.REPACK.2160p.PCOK.WEB-DL.DDP5.1.DV.HDR.H.265-FLUX`
File Name in NZB download: `The.Day.of.the.Jackal.S01E07.Episode.7.REPACK.2160p.PCOK.WEB-DL.DDP5.1.DV.HDR.H.265-FLUX.mkv`
File Name after rename during import: `The Day of the Jackal (2024) - S01E07 - Episode 7 [Peacock TV WEBDL-2160p Proper][DV HDR10][EAC3 5.1][x265]-FLUX.mkv`

As you can see, both the NZB name (Release Title) and the filename of the episode both contain `H.265`. Therefor, I suspect this is not the issue. After importing, the filename now includes `x265` (based on ffprobe metadata). 

However, this wouldn't account for the difference in the Custom Format Score anyway, since both `h265 (4k)` and `x265 (Web)` have a score of 0, which wouldn't affect the overall score. The only thing that could be affecting the score is the missing `Peacock TV` custom format after the file has been renamed. See the table below for the calculation.

| CFs of NZB | CFs after import & rename | Score in 2160p Optimal profile |
|------------|----------------------------------|----|
| `2160p`|`2160p`| 60 |
| `DD+` |`DD+` | 0 |
| `Dolby Vision` | `Dolby Vision` | 30 |
|`h265 (4k)` | `h265 (4k)` | 0 |
|`HDR10`| `HDR10` | 10 |
|`Peacock TV`| - | 20 |
| - |`x265 (Web)` | 0 |

Screenshot of the score of the file in the library:
<img width="1451" alt="image" src="https://github.com/user-attachments/assets/702e935a-b819-403d-8e9d-6435f0cdb6e9" />

I've checked the debug and trace logs. The upgrade decision making process determines that the score of the file that was previously downloaded and is currently in the library (same as what it intends to download again), is 20 points lower than the new release on the RSS feed (100 -> 120). 

This is what happens when you parse the NZB name and the filename after renaming in the library:

**NZB parsing:**
<img width="1053" alt="image" src="https://github.com/user-attachments/assets/41f7d753-fd6e-477e-90af-a110f35a2c6d" />
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/6e2df4af-80dc-4d69-96cf-f09323b0f62f" />

**File name parsing (renamed):**
<img width="1050" alt="image" src="https://github.com/user-attachments/assets/17eaceb9-40d7-4920-bf0a-1042b55630cd" />
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/0aad3c39-885f-4317-a54c-9d91407d6c15" />

After playing around and testing different strings for the filename, I've observed that the `Peacock TV` custom format is only applied when I remove the `x265` from the filename string. I've narrowed it down to this custom format having a negating Release Title condition with the regex `^(?!.*(?i:remux)).*([x]\s?(\.?265))`, which matches on the `x265` substring in the filename after importing.
<img width="860" alt="image" src="https://github.com/user-attachments/assets/1f583ad9-632f-4127-a5ee-c6ac3a0ce0c4" />

Any reason for having this `x265` negating condition in the "network" custom formats?

**Comment by santiagosayshey** [comment]
I can assure you that the file rename does NOT affect upgrade propensity. 

> Release Title - This is a regular expression matched against the release title and, after download, the filename on disk.
> Note: Sonarr only considers text after the series title which means anything preceding the title is ignored.
> Note that the matching text is needed to be included in the renamed file to ensure continual matching of the custom format after import.

This only talks about when the release title custom format applies, not what scores affect upgrades. You can rename all your files to whatever you want and it *will not* affect if something gets upgraded in the future. We have tested this thoroughly. 

The only time the custom format score matters is during a search, and on import. The score given to a rename is irrelevant. Not everybody uses the same rename philosophy. Personally, I only include title, year and resolution. If the rename score mattered, every single movie / tv show I've ever downloaded would get stuck in an infinite loop of redownloading the same thing over and over. This is the not the case.

Your observation about the x265 / peacock CFs is correct, but it doesn't affect anything. For the record, Peacock negates x265 so that encodes from a web-dl are not grabbed. 

What is the issue? Right now, I can't tell with the information I have. It almost certainly has to do with a file name mismatch on import. The torrent / file names you provided seem fine, but there could be issues with obfuscation or something else that we're missing. 

Could you please provide:

1. A history screenshot for your imports. Your should see something like this:

![image](https://github.com/user-attachments/assets/ed906abb-a73d-407f-b8ea-023d3605092e)

2. A link to the NZB(s) you have downloaded. 

Thanks
